### DISCUSSION TITLE
How to use mesa_geo and have the agents compute path and move on it?

### DISCUSSION THREAD

**Discussion by ghost** [mesa/mesa-geo#195] [Q&A] [answered]
I have a crowd-logistics model (created using mesa) in which I used continuous space to create my agents . These agents are supposed to move to a destination as needed in each tick. To make my model more realistic, I need to put these agents on an actual map and have them use actual routes. Is this possible?

I saw in another discussion on this group (using only routes #162) that referred to Agents and Networks example. I tried to down load this model but I get the following error:
**_ModuleNotFoundError: No module named 'src'_**

I have maintained the same folder structure as shown in the example.

**Comment by wang-boyu** [comment]
Sorry about the confusion. As mentioned in the [README](https://github.com/projectmesa/mesa-examples/blob/main/gis/agents_and_networks/README.md) file, in the Agents and Networks example folder, you'll need to do a

```bash
python3 -m pip install -r requirements.txt
```

so that the `src` module can be installed in the virtual environment (if you use a virtual environment). Then the model should be able to run or be imported.

Let me know if this works.

**Comment by ghost** [comment]
Hi @wang-boyu,

Thank you for the response. I followed the instructions. I used the command you showed above and installed the packages through the requirement.txt file. 

However, I still get thf following error:
```
usage: Agents and Networks in Python [-h] --campus CAMPUS
Agents and Networks in Python: error: the following arguments are required: --campus
An exception has occurred, use %tb to see the full traceback.

SystemExit: 2
```

**Comment by wang-boyu** [comment, marked as answer]
Hi, the model should be run via the following command (in the example directory):

```bash
python3 scripts/run.py --campus ub
```

or change `ub` to `gmu` for a different campus map:

```bash
python3 scripts/run.py --campus gmu
```

Hopefully this will get the model running : )

**Reply by ghost** [reply]
Hi,
Thanks for sharing this info. I was trying to run the run.py directly and that's why it was not working. 

It opens the interface now. 
![image](https://github.com/projectmesa/mesa-geo/assets/131771458/74997232-2d27-49ae-883f-e2d3ebc98ff1)

However, after the first step, it stops and gives the following error:

`!python3 scripts/run.py --campus ub
Interface starting at http://127.0.0.1:8521
Socket opened!
{"type":"reset"}
{"type":"reset"}
{"type":"get_step","step":1}
{"type":"get_step","step":2}
Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET', uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/Users/preetamkulkarni/anaconda3/lib/python3.11/site-packages/tornado/websocket.py", line 630, in _run_callback
    result = callback(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/preetamkulkarni/anaconda3/lib/python3.11/site-packages/mesa_viz_tornado/ModularVisualization.py", line 230, in on_message
    self.application.model.step()
  File "/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Python MESA/Agents and Networks/src/model/model.py", line 212, in step
    self.schedule.step()
  File "/Users/preetamkulkarni/anaconda3/lib/python3.11/site-packages/mesa/time.py", line 145, in step
    self.do_each("step", shuffle=True)
  File "/Users/preetamkulkarni/anaconda3/lib/python3.11/site-packages/mesa/time.py", line 121, in do_each
    getattr(self._agents[agent_key], method)()
  File "/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Python MESA/Agents and Networks/src/agent/commuter.py", line 88, in step
    self._prepare_to_move()
  File "/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Python MESA/Agents and Networks/src/agent/commuter.py", line 134, in _prepare_to_move
    self._path_select()
  File "/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Python MESA/Agents and Networks/src/agent/commuter.py", line 189, in _path_select
    self.model.walkway.cache_path(
  File "/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Python MESA/Agents and Networks/src/space/road_network.py", line 87, in cache_path
    with open(self._path_cache_result, "wb") as cached_result:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'outputs/ub_path_cache_result.pkl'`

Iget a similar error with "!python3 scripts/run.py --campus gmu"

**Reply by wang-boyu** [reply]
Yeah sorry this is a bug that just got fixed. Do you mind pulling latest code and try again?

**Reply by wang-boyu** [reply]
Or else you could also create an empty folder named `outputs` in the example folder. This should fix it.

**Comment by ghost** [comment]
Hi @wang-boyu,

Thank you for fixing the issue. It is working on my computer now. 

I have a crowd-logistics model (created using mesa) in which I used continuous space to create my agents . These agents are supposed to move to a destination as needed in each tick. To make my model more realistic, I need to put these agents on an actual map and have them use actual routes. Is this possible based on your experience with mesa-geo? I am trying to understand how hard this would be to achieve.

**Reply by wang-boyu** [reply]
Yes this is exactly what happens in the agents-and-networks example model. Commuters' [__path_select()](https://github.com/projectmesa/mesa-examples/blob/1e0d6b855c853c1dfa749348445a31000b4daca4/gis/agents_and_networks/src/agent/commuter.py#L177) function computer their shortest path on a road network, which is implemented as a separate class [here](https://github.com/projectmesa/mesa-examples/blob/main/gis/agents_and_networks/src/space/road_network.py). In this model all agents move at the same speed, and there's no agent-agent interactions (e.g., speed doesn't change if there's traffic jam ahead). You could add some heterogeneity by letting agents having different speed, whereas agent-agent interaction would require more effort to update the model.

**Reply by ghost** [reply]
Thank you!
